src/System.Linq/src/System/Linq/Union.cs

```diff
@@ -59,68 +83,52 @@ private bool GetNext()
                 return false;
             }
 
-            public override bool MoveNext()
+            public sealed override bool MoveNext()
             {
-                switch(state)
+                if (state == 1)
                 {
-                    case 1:
-                        _enumerator = _first.GetEnumerator();
-                        if (_enumerator.MoveNext())
-                        {
-                            state = 2;
-                        }
-                        else
+                    for (IEnumerable<TSource> enumerable = GetEnumerable(0); enumerable != null; enumerable = GetEnumerable(state - 1))
+                    {
+                        IEnumerator<TSource> enumerator = enumerable.GetEnumerator();
+                        ++state;
```

Comment:
Yeah, that'd be a good thing to fix.  We don't have tests for that?  I'm sure you'll be adding some :wink: